### DISCUSSION TITLE
Ingredient processor for non-english languages

### DISCUSSION THREAD

**Discussion by Lycolus** [mealie-recipes/mealie#1185] [General]
First of all, I only recently came upon this project and I must say I am really impressed!

I am currently running version v1.0.0b. As I am not from an English-speaking country the current version of the natural language processor is unfortunately not very helpful. Now I am aware that it is not easy to add another language so I plan to use the Brute processor. When trying to enter recipes I came upon some strange issues, however.

When experimenting with the parser under settings, entering the phrase:
'1 middelgroot ei' (hello fellow dutch people! English: '1 medium-sized egg'
Gets correctly split into:
Quantity: 1 Unit: Middelgroot Food: ei

However, when I add a recipe from a link, edit the ingredient list, and switch the parser to brute, then press 'parse all' the units disappear and I am left with just the quantity and food. From my experiment with the parser, I would conclude that the issue is not the parser itself. Am I missing something here or is this a bug in the beta release perhaps? (I noticed a small note saying 'add x food', but nothing similar for units)

Some related questions: Can I change the parser to brute by default instead of having to manually switch it every time?
Is there some way to get an overview of all units/foods, similar to what the tandoor app has for example? I have been pressing a lot of buttons but have been unable to find it so far.

Thanks for your time!

**Comment by zierbeek** [comment]
Hi fellow dutchman!

**Comment by hay-kot** [comment]
> However, when I add a recipe from a link, edit the ingredient list, and switch the parser to brute, then press 'parse all' the units disappear and I am left with just the quantity and food.

This is intended behavior. The parser only uses the note to parse ingredients out, so it's only expected that the note field is being used. There isn't currently a way to selectively parse ingredients. There may be in the future, but it isn't a priority for me at the moment. 

> Can I change the parser to brute by default instead of having to manually switch it every time?

Not at the moment. I'd be open to a PR to support this through and ENV variable though.

> Is there some way to get an overview of all units/foods, similar to what the tandoor app has for example? 

Yes, you can view all Foods and Units at `/group/data/foods` or through the user menu and then hit manage data.

---

In general I think the non-english experience is sub-par for a large portion of the application. I'm hoping someone from the community would be up for taking on some of the tasks to improve that experience in the application.

**Comment by Lycolus** [comment]
Thanks for the super quick reply @hay-kot  (And hallo @zierbeek !). I found the 'foods and units'.

> This is intended behavior. The parser only uses the note to parse ingredients out, so it's only expected that the note field is being used. There isn't currently a way to selectively parse ingredients. There may be in the future, but it isn't a priority for me at the moment.

On the first point, either I am misunderstanding your reply or I did not correctly explain my situation because this behavior feels very strange to me as intended behavior. At the very least I'll try to explain again using images.

So after importing with 'ingredients off' by default (I fear that is too much for my parents!) I can edit the recipe in order to turn ingredients on and end up with this situation:
![chrome_2022-05-01_20-20-13](https://user-images.githubusercontent.com/56881763/166159122-15d546f7-a1b7-47e7-b20e-387a883d1ac0.png)
As you can see, the quantity unit and food are all in the note field.

Next, I press the 'parse' button at the bottom of the ingredient list and switch the parser to brute and press parse all, which results in this:
![chrome_2022-05-01_20-29-09](https://user-images.githubusercontent.com/56881763/166159434-02eace82-ca1a-4bb5-a3d7-e60af0024d85.png)
I am assuming the text at the top is the input (the text in the note field) and as you can see it correctly detected the food 'ei' and quantity '1' but it completely discarded the 'middelgroot'. 

For reference when using the brute parser from 'settings' with the same input:
![chrome_2022-05-01_20-31-39](https://user-images.githubusercontent.com/56881763/166159520-720cb3df-3e2f-4fe2-ab76-292236782dcf.png)
You can see that it correctly detected the unit 'middelgroot'.

If you did in fact understand my situation correctly then I would like to kindly ask why you feel the current situation should be the intended behavior instead of adding the 'middelgroot' unit. Perhaps that will help me understand the design decision. Thanks again!

**Reply by hay-kot** [reply]
Ahhhh, I think I get it now. It doesn't prompt for units. If you don't have the unit added it won't do anything. It ONLY uses already created units to match against. Not ideal, perhaps there is some better behavior for this flow.

**Reply by Lycolus** [reply]
Perhaps someting akin to current system for foods?
e.g. when a 'food' is not recognized:
![chrome_2022-05-01_20-46-36](https://user-images.githubusercontent.com/56881763/166160078-7f2bab5c-8035-4808-a6e3-aae3eb585b99.png)
A second button could appear for units? Honestly I am not too well versed in programming UI elements so I am not sure how difficult this would be.

**Comment by bentolor** [comment]
Also need to chip into the hymn of praise: Really, really awesome project and work!

As a German user I'm to improve some of the rough edges for our daily usage. Therefore I'm currently trying to estimate what it would take, to provide natural language processors for other languages (i.e. German in my case) as well.

Yet I do understand, that the output in https://github.com/mealie-recipes/nlp-model is the result of some processing with https://github.com/mealie-recipes/crfpp on some dataset. But I was not able to spot the source/location of the "New York Times" dataset mentioned in https://nightly.mealie.io/contributors/guides/ingredient-parser/

I also see a poor-man hardcoded, language-specific preprocessing steps, which I think can be tackled easily.

Can you outline in a few bullet points the model creation aspect you applied so far for English? Thanks  lot!

**Comment by michael-genson** [comment]
There have been many improvements to the brute parser since this discussion was opened, so I'll go ahead and close it. FYI your last-used parser will be the default for you now as well
